x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -418,27 +476,146 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
+                        claimsSet.getIssueTime(),
+                        claimsSet.getNotBeforeTime(),
+                        claimsSet.getExpirationTime(),
+                        header.getKeyID(),
+                        claimsSet.getJWTID()
+                    );
+                }
+                // Validate all else before signature, because these checks are more helpful diagnostics than rejected signatures.
+                final boolean isJwtSigHmac = JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_HMAC.contains(header.getAlgorithm().getName());
+                JwtValidateUtil.validateType(jwt);
+                JwtValidateUtil.validateIssuer(jwt, allowedIssuer);
+                JwtValidateUtil.validateAudiences(jwt, allowedAudiences);
+                JwtValidateUtil.validateSignatureAlgorithm(jwt, isJwtSigHmac ? this.allowedJwksAlgsHmac : this.allowedJwksAlgsPkc);
+                JwtValidateUtil.validateAuthTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateIssuedAtTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateNotBeforeTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateExpiredTime(jwt, now, this.allowedClockSkew.seconds());
+
+                // At this point, client authc and JWT kty+alg+iss+aud+time filters passed. Do sig last, in case JWK reload is expensive.
+                try {
+                    JwtValidateUtil.validateSignature(
+                        jwt,
+                        isJwtSigHmac ? this.contentAndJwksAlgsHmac.jwksAlgs.jwks : this.contentAndJwksAlgsPkc.jwksAlgs.jwks
+                    );
+                } catch (Exception originalValidateSignatureException) {
+                    if (isJwtSigHmac) {
+                        throw originalValidateSignatureException; // HMAC reload not supported at this time
+                    }
+                    final String sigErr = originalValidateSignatureException.getMessage() + " ";
+
+                    ListenableFuture<Boolean> reloadAction = this.reloadActionRef.get(); // shared by threads using this realm
+                    final PlainActionFuture<Boolean> reloadListener = PlainActionFuture.newFuture(); // local thread
+                    final PlainActionFuture<Boolean> reloadListener2 = PlainActionFuture.newFuture(); // local thread
+                    while (reloadAction == null) {
+                        reloadAction = new ListenableFuture<>(); // current thread will try to take charge of reload
+                        boolean isCurrentThread = this.reloadActionRef.compareAndSet(null, reloadAction);
+                        if (isCurrentThread == false) {
+                            reloadAction = this.reloadActionRef.get(); // different thread took charge, get the shared reference
+                        }
+                        reloadAction.addListener(reloadListener);
```

Comment:
I copied that code from OIDC. I think it is a bug here in JWT realm, it is a bug there in OIDC realm too.

https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java#L959-L970

Reply:
I copied `future = reloadFutureRef.get();`  from OIDC, and renamed the variables to `reloadAction = this.reloadActionRef.get();`.

If it is problematic here in this JWT realm PR, then it could be problematic in OIDC realm too.

https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java#L959-L970

Reply:
No, OIDC's version is fine because the `addListener` call is guarded with `while (future == null) { ... }`. I don't think using `while` is great, but it does not have the same NPE issue.

Reply:
Replaced with Tim's proposed code. This should be resolved based on your comments about Tim's proposal.